age-piv-p256.md

```diff
@@ -0,0 +1,100 @@
+# age piv-p256 recipient stanza
+
+[c2sp.org/age-piv-p256](https://c2sp.org/age-piv-p256)
+
+## Introduction
+
+This document specifies the age piv-p256 recipient stanza. This stanza
+is used for encryption using NIST P-256 keys stored on hardware tokens
+or PIV devices.
+
+Apart from using the NIST P-256 curve, the piv-p256 recipient stanza
+differs from the native [X25519 recipient
+stanza](https://age-encryption.org/v1#x25519-recipient-stanza) in that
+it includes a tag of the recipient, to allow the recipient to select the
+correct secret without the hardware token requesting user interaction.
+
+## Conventions used in this document
+
+ABNF syntax follows [RFC
+5234](https://www.rfc-editor.org/rfc/rfc5234.html) and [RFC
+7405](https://www.rfc-editor.org/rfc/rfc7405.html) and references the
+core rules in RFC 5234, Appendix B.1.
+
+The base64 encoding used throughout is the standard Base 64 encoding
+specified in [RFC 4648](https://www.rfc-editor.org/rfc/rfc4648.html),
+Section 4, without `=` padding characters (sometimes referred to as
+"raw" or "unpadded" base64). Encoders MUST generate canonical base64
+according to RFC 4648, Section 3.5, and decoders MUST reject
+non-canonical encodings and encodings ending with `=` padding
+characters.
+
+`||` denotes concatenation. `0x` followed by two hexadecimal characters
+denotes a byte value in the 0-255 range.
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
+"OPTIONAL" in this document are to be interpreted as described in [BCP
+14](https://www.rfc-editor.org/info/bcp14) [RFC
+2119](https://www.rfc-editor.org/rfc/rfc2119.html) [RFC
+8174](https://www.rfc-editor.org/rfc/rfc8174.html) when, and only when,
+they appear in all capitals, as shown here.
+
+## Recipient stanza
+
+A piv-p256 recipient stanza has three arguments.
+
+    -> piv-p256 8u8pZg Aqa5k2Tm/BJGTzCno1W29dWaWBHkFchOztcIZ7aQo0Ca
+    CXE0N0KMI0BUE5WaYTLOxYv7aHB/5IQpAZKsk2C/yjU
+
+1.  The first argument is the fixed string `piv-p256`.
+
+2.  The second argument is the base-64 encdoded tag of the recipient,
+    consisting of the first 4 bytes of the SHA-256 hash of the
+    recipient:
+
+         tag = SHA-256(recipient)[0..4]
+
+3.  The third argument is the base64-encoded ephemeral share computed by
+    the recipient implementation as follows:
+
+         ephemeral secret = read(CSPRNG, 32)
+         ephemeral share = P256(ephemeral secret, basepoint)
```

Comment:
"P256" is not a well-defined ECDH operation the way X25519 is.

We should define it somewhere as scalar multiplication over P-256, with SEC 1 point decoding and encoding. (Assuming that's what we are doing? But then why is the third argument 32 bytes?)

Reply:
My comments about using the x-coordinate for ECDH with P-256 (and really all the NIST P-curves) were supposed to be in reply to this comment. I guess I must've clicked the wrong one in my previous replies.

Reply:
Replying to https://github.com/C2SP/C2SP/pull/31#discussion_r1305720851.

The ECDH output is the x-coordinate. The ECDH input (that is, the third stanza argument here) is `0x04 || x || y`, which is SEC 1. 

https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.8.2